src/vsr/journal.zig

```diff
@@ -971,11 +1649,20 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                 return;
             }
 
+            const slot = self.slot_with_header(message.header).?;
+            if (!self.prepare_inhabited[slot.index] or
+                self.prepare_checksums[slot.index] != message.header.checksum)
+            {
+                self.write_prepare_debug(message.header, "entry changed then restored while writing headers");
+                self.write_prepare_release(write, null);
+                return;
+            }
+
```

Comment:
> Perhaps what we should rather do here is, if we see that we still have the header when calling `has()` on line 1646, and then see that the `prepare_checksum` was marked uninhabited (which must have happened between writing the prepare and writing the header), then re-assign `prepare_checksum` here again, asserting that the slot is dirty.

I think that would work, but I don't understand why it's preferable to either the implemented solution, or to simply not modifying `prepare_{inhabited,checksums}` in `remove_entries_from`. (The latter solution might actually be best, since `prepare_checksums`' original purpose was to directly reflect the WAL's prepare ring on disk. And `remove_entries_from` doesn't actually overwrite anything on disk, so the prepare is still intact.)

> On a related note, I think we might want `has_clean()` to assert that the `prepare_checksum` also matches?

:+1:

Reply:
> I don't understand why it's preferable

When I did the `remove_entries_from()` fix, I was only thinking of keeping things in sync. Since the log truncation usage of `remove_entries_from()` is semantically about removing ops, I thought that that should apply to `prepare_inhabited` too, or it might surprise the caller, in the same way that we hit that assertion failure on the dirty bit.

Reinstating the `prepare_checksum` here could also help show awareness about this particular scenario, in the right place close to where the concurrency issue happens.

However:

> The latter solution might actually be best, since prepare_checksums' original purpose was to directly reflect the WAL's prepare ring on disk

Let's do that rather then. This is simpler, and we can comment in `remove_entries_from()` why we're careful to leave it untouched.